### PR TITLE
Feature/wdboggs/timeformat iso8601

### PR DESCRIPTION
**Pull request by darianboggs**
## Description

Implementing ISO 8601 (Date/Time) support for MAPL
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [ ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [ ] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [ ] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: base/MAPL_ISO8601_Time.F90 [hunk 1]
```diff
@@ -0,0 +1,1292 @@
...
+      integer :: timezone_offset_
+      logical :: is_valid_ = .FALSE.
+   end type time_fields
+
+   interface time_fields
+      module procedure :: construct_time_fields
+   end interface time_fields
+
+   type :: ISO8601DateTime
+#ifndef DEBUG
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:173]: Please eliminate the `#ifndef DEBUG` prior to merge.

File: base/MAPL_ISO8601_Time.F90 [hunk 2]
```diff
@@ -0,0 +1,1292 @@
...
+      private
+#endif
+      type(ISO8601Date) :: date_
+      type(ISO8601Time) :: time_
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:177]: Looks like you probably do not need the trailing underscores now, as your accessors all have `get_` ...

File: base/MAPL_ISO8601_Time.F90 [hunk 3]
```diff
@@ -0,0 +1,1292 @@
...
+         c = isostring(pos:pos)
+         if(time_found) then
+            select case(c) 
+               case('H')
+                  if(hours >= 0 .or. minutes >= 0 .or. &
+                     seconds >= 0 .or. istart < 1) cycle
+                  hours = read_whole_number_indexed(isostring, istart, istop)
+                  if(hours < 0) cycle
+                  duration%hours_= hours ! DEBUG WDB
+                  istart = 0
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:714]: There seems to be lots of problems here.  First, the entire conditional is very long and screams for helper procedures.

But it is also the case that `hours` is a saved variable and would seem to accidentally be positive the second time this procedure is called.    I at least cannot follow the logic.   

Also, istart is given a value at the end of this subblock, and then never referenced again ...

File: base/MAPL_ISO8601_Time.F90 [hunk 4]
```diff
@@ -0,0 +1,1292 @@
...
+         successful = .TRUE.
+      end do
+
+      if(successful) then
+!         duration%years_= years
+!         duration%months_= months
+!         duration%days_= days
+!         duration%hours_= hours
+!         duration%minutes_= minutes
+!         duration%seconds_= seconds
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:779]: Uncomment or delete.

File: base/MAPL_ISO8601_Time.F90 [hunk 5]
```diff
@@ -0,0 +1,1292 @@
...
+
+   ! start_date stop_date
+   ! duration stop_date
+   ! start_date duration
+   function construct_ISO8601Interval(isostring, rc) result(interval)
+      character(len=*), intent(in) :: isostring
+      integer, intent(inout) :: rc
+      type(ISO8601Interval) :: interval
+      integer :: status
+      _RETURN(_FAILURE)
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:794]:
```suggestion
      _FAIL('Not implemented')
      _RETURN(_SUCCESS)
```

File: base/MAPL_ISO8601_Time.F90 [hunk 6]
```diff
@@ -0,0 +1,1292 @@
...
+!      do i=1,len(digit_string)
+!         if(is_digit(string(i:i))) then
+!            j = j + 1
+!            digit_string(j:j) = string(i:i)
+!         end if
+!      end do
+!
+!      digit_string = trim(digit_string)
+!
+!   end function get_digits
```
Comment (tclune) [inline base/MAPL_ISO8601_Time.F90:1069]: Uncomment or delete.

**Review by tclune — CHANGES_REQUESTED** [review]
See inline comments.

File: shared/tests/test_MAPL_ISO8601_DateTime.pf [hunk 7]
```diff
@@ -635,21 +636,21 @@ contains
...
+!      @assertEqual(513, time%millisecond_)
+!      @assertEqual(0, time%timezone_offset_)
+
+!      time = parse_time('T174107.513-0500')
+!      @assertTrue(time%is_valid_)
+!      @assertEqual(17, time%hour_)
+!      @assertEqual(41, time%minute_)
+!      @assertEqual(7, time%second_)
+!      @assertEqual(513, time%millisecond_)
+!      @assertEqual(-300, time%timezone_offset_)
```
Comment (tclune) [inline shared/tests/test_MAPL_ISO8601_DateTime.pf:653]: Now the tests have comments that should either be deleted or uncommented ...

File: shared/MAPL_ISO8601_DateTime.F90 [hunk 8]
```diff
@@ -710,22 +709,19 @@ function construct_ISO8601Duration(isostring, imin, imax, rc) result(duration)
...
                   if(minutes < 0) cycle
                   if(hours < 0) hours = 0
-                  duration%minutes_= minutes ! DEBUG WDB
                   istart = 0
                case('S')
                   if(seconds >= 0 .or. istart < 1) cycle
                   seconds = read_whole_number_indexed(isostring, istart, istop)
                   if(seconds < 0) cycle
                   if(hours < 0) hours = 0
                   if(minutes < 0) minutes = 0
```
Comment (tclune) [inline shared/MAPL_ISO8601_DateTime.F90:724]: Still lots of references to uninitialized values scattered through this code.

**Review by tclune — DISMISSED** [review]
See inline comments.

File: base/MAPL_ISO8601_Time.F90 (hunk 4 above)
Comment (darianboggs) [reply base/MAPL_ISO8601_Time.F90]: Fixed

File: base/MAPL_ISO8601_Time.F90 (hunk 5 above)
Comment (darianboggs) [reply base/MAPL_ISO8601_Time.F90]: Replaced.

File: base/MAPL_ISO8601_Time.F90 (hunk 6 above)
Comment (darianboggs) [reply base/MAPL_ISO8601_Time.F90]: Deleted

**Comment by darianboggs** [comment]
Clean up for merge

**Review by mathomp4 — APPROVED** [review]

**Pull request merged**